User management (Devise, model, controller set-up)

robot269006 opened the pull request:
# WHAT
Introduce devise gem and apply user registrations.
- [x] ### 1. Install devise gem
- [x] ### 2. Download materials
- [x] ### 3. Add view files for devise
- [x] ### 4. Swap devise view with downloaded materials
- [x] ### 5. Create users model
- [x] ### 6. Create users_controller
- [x] ### 7. Add sign-up function

# WHY
In order to make preparations for user management

robot269006 committed:
Devise install

1. gem
2. bundle install
3. rails g devise:install

robot269006 committed:
View install & relocate

1. Download view materials
2. rails g devise:views
3. Relocate view files

robot269006 committed:
Create User model

1. Delete User table that was created by mistake
    A. rails destroy model User
    B. rails g migration DeleteUsersModel
    C. drop_table :users
    D. bundle exec rake db:migrate
2. rails g devise User
3. Check migration file (No changes edited)
4. bundle exec rake db:migrate

robot269006 committed:
Create user controller

1. Create user controller by rails g controller user
2. Moved user.scss file to views folder

robot269006 committed:
Registrations implementation

Registrations page defines name (“= f.text_field :name, autofocus:
true” under file registrations/new.html.haml) in order to enable name
registration, but currently there’s no column that apply. Therefore,
edit migration file to add name column

1. rails g migration AddNameToUsers
2. Edit migration file => add_column :users, :name, :string, null:false
3. bundle exec rake db:migrate

robot269006 committed:
Convert all erb to haml

1. rake -T => check the command to replace erb files
2. rake haml:replace_erbs

robot269006 committed:
SCSS and HAML adjustments

1. import user.scss
2. convert erb files to haml

FumiyaMori commented on app/assets/stylesheets/application.scss:
```diff
@@ -14,3 +14,4 @@
 @import "button.scss";
 @import "icon.scss";
 @import "body.scss";
+@import "user.scss";
```
ファイルの最後には改行を入れましょう

robot269006 replied:
対応しました

FumiyaMori commented on app/controllers/user_controller.rb:
```diff
@@ -0,0 +1,2 @@
+class UserController < ApplicationController
```
Usersのsが抜けてしまっています。正しくは`UsersController`かと思います。
UsersControllerはまだ使わないので、このコントローラは消してしまい必要になったときに正しく生成し直すかたちで良いかと思います。

robot269006 replied:
対応しました

FumiyaMori commented on app/views/devise/edit.html.haml:
```diff
@@ -0,0 +1,39 @@
+#account-page.account-page
```
このファイルの配置場所は正しいでしょうか？
もう一度確認して見てください。

robot269006 replied:
/devise/sessionsの中に配置しました

FumiyaMori commented on app/views/user/edit.html.haml:
```diff
@@ -0,0 +1,39 @@
+#account-page.account-page
```
chatspaceではdeviseの機能を使いUserの編集を行うのでこのファイルは必要ありません
アカウントの編集画面用のビューファイルをdeviseの正しいディレクトリに配置できれば問題ないです

robot269006 replied:
上記の通り/devise/sessions内に配置しました

FumiyaMori commented on db/migrate/20170408080550_add_name_to_users.rb:
```diff
@@ -0,0 +1,5 @@
+class AddNameToUsers < ActiveRecord::Migration[5.0]
+  def change
+    add_column :users, :name, :string, null:false
```
nameカラムにはindexを付け、ユニーク制約を付けたほうが良いかと思います

robot269006 replied:
migrationファイルを作成し、add_indexでインデックス付与とユニーク化を実行しました

robot269006 committed:
Revision

1. A blank space at the end of application.scss
2. rails destroy controller User -> destroy controller
3. Moved /devise/edit.html.haml to /devise/sessions/edit.html.haml

robot269006 committed:
Revision

1. added space in application.scss
2. rails destroy controller User
3. moved /devise/edit.html.haml to /devise/sessions/edit.html.haml
4. rails g migration AddIndexToUsers
5. edited migration file add_index :users, :name, unique:true
6. bundle exec rake db:migrate

robot269006 committed:
Merge remote-tracking branch 'origin/user-management' into user-management

KoichiOzaki commented on app/assets/stylesheets/user.scss:
```diff
@@ -0,0 +1,141 @@
+.account-page {
+  .account-page__inner {
```
sassを使っているのであれば&記法を使用しましょう。
親要素の名前を継承可能なはずです。
これ以降の記述でも&記法が使用可能な場所を修正してみましょう。

robot269006 replied:
修正しました。以降、他に記法が適用できる所はしていくようにします。

KoichiOzaki commented:
大きな機能単位の追加をして、画面レイアウトを変更した場合はスクリーンショットを貼っていただけると助かります。

robot269006 commented:
<img width="1237" alt="2017-04-11 20 55 07" src="https://cloud.githubusercontent.com/assets/24761206/24908063/a46fdc72-1ef9-11e7-9fd7-19eba15a1401.png">

robot269006 committed:
Revision

1. Sass markup on user.scss

KoichiOzaki commented on app/models/user.rb:
```diff
@@ -0,0 +1,6 @@
+  # Include default devise modules. Others available are:
+  # :confirmable, :lockable, :timeoutable and :omniauthable
+  devise :database_authenticatable, :registerable,
+         :recoverable, :rememberable, :trackable, :validatable
```
nameカラムにデータはうまく入っているでしょうか？
サインアップ、サインイン、サインアウト、の機能を一通り動かしてみて問題なく動きそうか確認していただけると良いかと思います。

robot269006 replied:
@KoichiOzaki
確かにnameにdefault valueが設定されてません、みたいなメッセージがでますね。
少し調べてまたご連絡します！

robot269006 replied:
@KoichiOzaki
原因として、①nameカラムのdefaultを""に設定する　②null: trueにしてしまう　③MySQLのstrictモードの設定をオフにする
上記の３パターンで解決しようと考えているのですが、どれが正解なのかわかりません。。。
ご教示いただけないでしょうか？

KoichiOzaki replied:
そうですね、結論から言うとそれとは別のアプローチを取ったほうが良さそうです。
よくシステムのエラーを追っていると忘れがちになってしまいますが、本来ならどういう動作をすべきか（機能レベルで）というところにぜひ立ち返ってください。
今回は
機能１．ユーザ登録時に名前を入力でき、DBにデータを登録できること。
機能２．ユーザ登録時に名前を未入力の場合にエラーが発生すること。
という機能を実装したいので、
①、②、③のどの対応でもこの機能２から遠ざかってしまいそうです。
ということは別のアプローチが必要ですし、エラーの原因も別にありそうです。

KoichiOzaki replied:
deviseでサインアップ時にemail以外のカラムへデータを保存する方法
について調べてみると良いかと思います。

robot269006 replied:
確かにエラー回避に集中してしまってました。。本来の目的からずれてますね。
ご指摘の内容を少し探してみまして、application_controller.rbでpermit部分が実装されていないことに要因がありそうなことが分かったので、そちらの修正を行いたいと思います。
完了次第、また追ってご連絡させていただきます！

robot269006 committed:
Revision (allow registration)

allowed name value to be registered when signing up by
adding below to application_controller.rb

before_action :configure_permitted_parameters, if: :devise_controller?

protected
def configure_permitted_parameters
    devise_parameter_sanitizer.permit(:sign_up, keys: [:name])
end

robot269006 commented:
先ほどsign_up時にnameを登録許可できるように修正しました。
rails sで起動し、登録可能なことは確認済みです。
